### PR TITLE
Fix ArviZ version on test environment

### LINKED ISSUE #4034: Tests fail with: TypeError: from_pymc3() got an unexpected keyword argument 'save_warmup'
Seems like an arviz incompatibility. CC @AlexAndorra

### PR DESCRIPTION
**Pull request by AlexAndorra**
Related to #4034 
- Removed main conda channel from environment-dev.yml
- Added `dill` as pip dependency in environment-dev.yml
- Added kwarg to pytest fixture

### CODE REVIEW

**Comment by twiecki** [comment]
Looks like there a few more fixture errors:
```
==================================== ERRORS ====================================

__________ ERROR collecting pymc3/tests/test_variational_inference.py __________

pymc3/tests/test_variational_inference.py:578: in <module>

    ids=['mini', 'full']

../../../miniconda3/envs/testenv/lib/python3.6/site-packages/_pytest/fixtures.py:1358: in fixture

    warnings.warn(FIXTURE_POSITIONAL_ARGUMENTS, stacklevel=2)

E   pytest.PytestDeprecationWarning: Passing arguments to pytest.fixture() as positional arguments is deprecated - pass them as a keyword argument instead.

------------------------------- Captured stderr --------------------------------
```

Also seems like mkl is not being installed: `WARNING (theano.configdefaults): install mkl with `conda install mkl-service`: No module named 'mkl'`

**Comment by AlexAndorra** [comment]
> Looks like there a few more fixture errors:

Yep, I'm on it

> Also seems like mkl is not being installed:

Do you know what this is used for? Should we install with pip instead?

**Comment by twiecki** [comment]
@AlexAndorra MKL is used to speed up theano, it's optional but should get pulled in automatically so I'm a bit surprised it isn't. It's easiest to install via conda. We can just leave that for now.

**Comment by twiecki** [comment]
We're still getting the old arviz :-/

**Comment by AlexAndorra** [comment]
Even worse is that Travis still doesn't find ArviZ 🤦 You encountered this error the other day
```
PackagesNotFoundError: The following packages are missing from the target environment:
  - arviz
```

**Comment by AlexAndorra** [comment]
Ok, I'm officially out of ideas, at least for today 😩 
It seems that `The environment is inconsistent, please check the package plan carefully` (see [here](https://travis-ci.org/github/pymc-devs/pymc3/jobs/713308072#L1391)):
```
The following packages are causing the inconsistency:
  - conda-forge/linux-64::libblas==3.8.0=17_openblas
  - conda-forge/linux-64::xorg-libxt==1.1.5=h516909a_1003
  - conda-forge/linux-64::libgomp==9.3.0=h24d8f2e_11
  - conda-forge/linux-64::_openmp_mutex==4.5=0_gnu
  - conda-forge/linux-64::liblapack==3.8.0=17_openblas
  - conda-forge/linux-64::xorg-libxpm==3.5.13=h516909a_0
  - conda-forge/linux-64::libcblas==3.8.0=17_openblas
  - conda-forge/linux-64::xorg-libsm==1.2.3=h84519dc_1000
```
 I don't really know what this means though. I think that somehow it doesn't update our test env and that's why we get stuck with the old ArviZ version 🤷‍♂️ Any ideas?

**Comment by aseyboldt** [comment]
Why remove `defaults`? They never do that in the conda-forge docs (eg https://conda-forge.org/) as far as I know.
About mkl: We should install it as `"blas=*=mkl"` (see https://conda-forge.org/docs/maintainer/knowledge_base.html#blas). I think we should then also install mkl-service, as an interface to mkl for theano, but I'm not sure if this is actually needed or even a good idea.
Maybe the missing `defaults` in combination with the cache is messing things up now?

**Comment by twiecki** [comment]
@aseyboldt The default channel has an old arviz version that is somehow getting pulled in.

**Comment by AlexAndorra** [comment]
Yep, the problem is that we can't figure out how to not pull in the old
ArviZ that's living on the default channel. Even removing this channel from
the requirements file (which I agree isn't good practice) doesn't seem to
have any effect

Le ven. 31 juil. 2020 à 10:11, Thomas Wiecki <notifications@github.com> a
écrit :

**Comment by twiecki** [comment]
I actually do think it's good practice. conda-forge seems better and more-up-to-date in every way and mixing the two only causes problems.

**Comment by aseyboldt** [comment]
About `defaults`: In the conda-forge docs they never disable the defaults channel, only ever add conda-forge with higher priority and strict priority setting. The note here says they are still using packages from defaults: https://conda-forge.org/docs/user/tipsandtricks.html#how-to-fix-it. Could be out of date I suppose, but I haven't seen anything that would indicate that.
I think the mixing problems only happen (should only happen?) if the priority isn't set to strict.
As far as I can see it does use the most recent arviz now. The test failure doesn't seem related to this.

**Comment by twiecki** [comment]
Oh, cool so we solved it. I'm not strongly opinionated on the default channel, this solution seems fine.

Last error, which seems like a precision problem:
```
=================================== FAILURES ===================================

____________________________ test_fit_oo[SVGD-mini] ____________________________

inference = <pymc3.variational.inference.SVGD object at 0x7fb91b5bbb38>

fit_kwargs = {'n': 300, 'obj_optimizer': functools.partial(<function adagrad_window at 0x7fb9532831e0>, n_win=7, epsilon=0.1, learning_rate=0.075)}

simple_model_data = {'d': 111.36111111111111, 'data': Minibatch, 'mu0': 4.0, 'mu_post': -4.921929491177873, ...}

    @pytest.mark.run('first')

    def test_fit_oo(inference,

                    fit_kwargs,

                    simple_model_data):

        trace = inference.fit(**fit_kwargs).sample(10000)

        mu_post = simple_model_data['mu_post']

        d = simple_model_data['d']

        np.testing.assert_allclose(np.mean(trace['mu']), mu_post, rtol=0.05)

>       np.testing.assert_allclose(np.std(trace['mu']), np.sqrt(1. / d), rtol=0.1)

E       AssertionError: 

E       Not equal to tolerance rtol=0.1, atol=0

E       

E       Mismatched elements: 1 / 1 (100%)

E       Max absolute difference: 0.01228255

E       Max relative difference: 0.12961506

E        x: array(0.107044)

E        y: array(0.094762)

pymc3/tests/test_variational_inference.py:712: AssertionError
```

CC @ferrine

**Comment by AlexAndorra** [comment]
> Oh, cool so we solved it

Ha ha, it looks like it! This is weird, since we tested with the default channel yesterday 🤷‍♂️ Debugging Travis is more an more mysterious for me...

In addition to the error Thomas spotted, there is also this strange one:
`The command "eval git clone --depth=50 --branch=pull/4037 https://github.com/pymc-devs/pymc3.git pymc-devs/pymc3 " failed 3 times.`. 
May it be just a time out issue that'll disappear next time tests are run?

**Comment by aseyboldt** [comment]
I think the git thing should go away on its own.
The caches often make travis much more unreliable. There might also be a bug with the env there somehow. You can delete them to start fresh.

**Comment by aseyboldt** [comment]
Could one of these be the reason it failed?
https://github.com/conda/conda/issues/7238
https://github.com/conda/conda/issues/9065
If the order of channels isn't respected by `conda env create`, then we might accidentally end up with a higher priority on `defaults` from time to time.

**Comment by AlexAndorra** [comment]
The first one does look a lot like the problem we're having. If that is indeed the issue, the thing that's bugging me is that this behavior doesn't seem random: all our attempts have failed updating ArviZ through conda-forge instead of conda since yesterday 🤔

**Comment by AlexAndorra** [comment]
Wondering whether adding conda-forge channel in `created_testenv.sh` in addition to `environement-dev.yml` creates conflict, so let's see 🤷‍♂️

**Comment by AlexAndorra** [comment]
And... this doesn't work 🎉 --> Reverting

**Comment by ferrine** [comment]
Hmm

**Comment by ericmjl** [comment]
@AlexAndorra, I just tried creating the environment on my home GPU tower.

Instead of using conda, I used mamba. Confirmed that ArviZ 0.9 was installed in a clean new env.

Try this in `create_testenv.sh`.

```bash
if [ -z ${GLOBAL} ]; then
  if conda env list | grep -q ${ENVNAME}; then
    echo "Environment ${ENVNAME} already exists, keeping up to date"
  else
    conda config --add channels conda-forge
    conda config --set channel_priority strict
    conda install -c conda-forge mamba
    mamba env create -f environment-dev.yml
  fi
  source activate ${ENVNAME}
fi
```

**Comment by ferrine** [comment]
I've checked the branch locally. there is one failing test that should be fixed

```
+++ b/pymc3/tests/test_variational_inference.py
@@ -980,10 +980,10 @@ def test_var_replacement():
 def test_empirical_from_trace(another_simple_model):
     with another_simple_model:
         step = pm.Metropolis()
-        trace = pm.sample(100, step=step, chains=1)
+        trace = pm.sample(100, step=step, chains=1, tune=0)
         emp = Empirical(trace)
         assert emp.histogram.shape[0].eval() == 100
-        trace = pm.sample(100, step=step, chains=4)
+        trace = pm.sample(100, step=step, chains=4, tune=0)
         emp = Empirical(trace)
         assert emp.histogram.shape[0].eval() == 400
```

**Comment by canyon289** [comment]
Reached out for help upgrading the default in default anaconda channel
https://github.com/ContinuumIO/anaconda-issues/issues/11973

**Comment by ericmjl** [comment]
@AlexAndorra now try removing the `conda update --all` command.

**Comment by AlexAndorra** [comment]
Almost all is green now (so, kind of 🎉 I'd say), but we still have a `PackagesNotFoundError` error with ArviZ [during the last job](https://travis-ci.org/github/pymc-devs/pymc3/jobs/715919851#L1466) ("ArviZ compatibility"), and a [tolerance error in VI test](https://travis-ci.org/github/pymc-devs/pymc3/jobs/715919850#L4468) -- @ferrine, can I use a higher tolerance than 0.1 for this test?

**Comment by kyleabeauchamp** [comment]
Build is green!!!

**Comment by AlexAndorra** [comment]
OH MY GOD, my latest attempt worked: tests are now passing!! 🍾 
@twiecki, once you review and (hopefully!) approve, we can merge and, above all, go on merging the other poor PRs which are waiting to get merged 🥳

**Review by kyleabeauchamp — APPROVED** [review]
LGTM, but one question: if we need to use mamda to get our CI/CD working, does that mean our users will also be hit with the same bug if they don't use mamba?

**Comment by AlexAndorra** [comment]
IIUC, I don't think so: we used it here because it was one of the things we hadn't tried to debug esoteric and hard-to-understand Travis errors. Everything should work fine on other installations (I haven't experienced or heard of any similar troubles) -- cc @ericmjl

**Comment by ericmjl** [comment]
@kyleabeauchamp I suggested mamba because:

1. It's faster that `conda`,
2. It's a drop-in replacement for `conda`,
3. It will thus speed up the builds, at least the environment creation step.

In principle, we can change this back to `conda` if really desired. My diagnosis of why the problem existed, though, is because of the `conda update --all` command that was in the test env script. Don't know why it was there, but in retrospect, I don't think should have been in the build. Trying to update everything meant we'd run into dependency solver issues at some point, I think. (Just my analysis based on debugging while _also_ trying to listen in on the dev team call!)

File: pymc3/tests/test_variational_inference.py
Comment (ferrine) [inline pymc3/tests/test_variational_inference.py:712]: Dirty but ok, I think this might be related to some random stuff

Comment (ferrine) [reply pymc3/tests/test_variational_inference.py:712]: The thing is I did not have this error locally

Comment (AlexAndorra) [reply pymc3/tests/test_variational_inference.py:712]: Ha ha yeah, I was determined to make this test pass 😅 If this is fine by you, I can go ahead and merge then 🎉 

> I think this might be related to some random stuff

I agree, which is probably why you didn't see it locally 🤷‍♂️

**Comment by twiecki** [comment]
You're my hero @AlexAndorra <3 ;).

**Comment by ericmjl** [comment]
Woot! Great work, Alex!!

**Comment by AlexAndorra** [comment]
Ha ha ha thanks guys, and thanks for guidance to get it to the finish line Eric 😉

**Comment by jankaWIS** [comment]
Hi, I'm getting this error and seems to be related:
```python
TypeError                                 Traceback (most recent call last)
<ipython-input-21-b4e1b40cb924> in <module>
     12 #     trace_test = pm.sample(1000, tune=2000, random_seed=RANDOM_SEED)
     13 #     trace_test = pm.sample(random_seed=RANDOM_SEED)
---> 14     trace_test = pm.sample()
     15     test_prior = pm.sample_prior_predictive()

~/anaconda3/lib/python3.8/site-packages/pymc3/sampling.py in sample(draws, step, init, n_init, start, trace, chain_idx, chains, cores, tune, progressbar, model, random_seed, discard_tuned_samples, compute_convergence_checks, callback, return_inferencedata, idata_kwargs, mp_ctx, pickle_backend, **kwargs)
    622         if idata_kwargs:
    623             ikwargs.update(idata_kwargs)
--> 624         idata = arviz.from_pymc3(trace, **ikwargs)
    625 
    626     if compute_convergence_checks:

TypeError: from_pymc3() got an unexpected keyword argument 'save_warmup'
```

when running just a simple model:
```python
import arviz as az
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
import pymc3 as pm

RANDOM_SEED = 123

a = 5
k1 = 1.3
x_test = np.linspace(1,10,20)

y_test = a*np.exp(-k1*x_test)

# fit the model:
with pm.Model() as test_exp:
    a = pm.Normal("a", 10, 3)
    k1 = pm.Lognormal("k1", 0.5, 0.2)
    
    lam = a * pm.math.exp(-k1 * x_test)

    obs = pm.Poisson("y", lam, observed=y_test)

#     trace_test = pm.sample(1000, tune=2000, random_seed=RANDOM_SEED)
#     trace_test = pm.sample(random_seed=RANDOM_SEED)
    trace_test = pm.sample()
    test_prior = pm.sample_prior_predictive()

```

Why is that, how to fix it and what is going on?
Thanks

**Comment by OriolAbril** [comment]
@jankaWIS it looks like you have a very recent pymc3 version installed together with an old arviz version, updating arviz should fix it, however I'd recommend updating both.

Side note, pymc3 requires an arviz version recent enough for this to not happen as far as I remember so the root cause of this could be a broken environment where due to both lower and upper bounds it becomes impossible to comply with all the requirements

**Comment by jankaWIS** [comment]
I was using `'0.6.1'` and for some reason that was what `pip install arviz` gave me... Anyway, `conda install -c conda-forge arviz` installed the latest `0.10.0` version and that fixed the problem. Thanks! I wasn't aware of that.

**Comment by OriolAbril** [comment]
I am guessing you somehow had 0.6.1 installed, thus `pip install arviz` does nothing, it's already installed. In these cases you need to use `pip install -U arviz`, the u flag standing for upgrade

**Pull request merged**
